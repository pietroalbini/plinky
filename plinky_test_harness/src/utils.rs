use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
