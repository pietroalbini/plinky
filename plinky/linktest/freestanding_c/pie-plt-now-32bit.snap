---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: calculated layout
 │
 │    Layout:
 │  ╭─────────────────────────────┬─────────────┬─────────────┬────────────────┬───────────────╮
 │  │ Part                        │ File offset │ File length │ Memory address │ Memory length │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <elf header>                │ 0x0         │ 0x34        │ 0x0            │ 0x34          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <program header>            │ 0x34        │ 0x120       │ 0x34           │ 0x120         │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynstr                     │ 0x154       │ 0xd         │ 0x154          │ 0xd           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynsym                     │ 0x161       │ 0x30        │ 0x161          │ 0x30          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .hash                       │ 0x191       │ 0x18        │ 0x191          │ 0x18          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .dynamic                    │ 0x1a9       │ 0x60        │ 0x1a9          │ 0x60          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rel.plt                    │ 0x209       │ 0x10        │ 0x209          │ 0x10          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .interp                     │ 0x219       │ 0x55        │ 0x219          │ 0x55          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <section header>            │ 0x26e       │ 0x2f8       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .shstrtab                   │ 0x566       │ 0xa1        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .strtab                     │ 0x607       │ 0x51        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .symtab                     │ 0x658       │ 0x80        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .comment#0                  │ 0x6d8       │ 0x22        │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x6fa       │ 0x906       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .rodata                     │ 0x1000      │ 0xd         │ 0x1000         │ 0xd           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .eh_frame                   │ 0x100d      │ 0x48        │ 0x100d         │ 0x48          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x1055      │ 0xfab       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#0                     │ 0x2000      │ 0x32        │ 0x2000         │ 0x32          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text.__x86.get_pc_thunk.bx │ 0x2032      │ 0x4         │ 0x2032         │ 0x4           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .text#1                     │ 0x2036      │ 0x25        │ 0x2036         │ 0x25          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .plt                        │ 0x205b      │ 0x30        │ 0x205b         │ 0x30          │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ <padding>                   │ 0x208b      │ 0xf75       │ -              │ -             │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .got                        │ 0x3000      │ 0x0         │ 0x3000         │ 0x0           │
 │  ├─────────────────────────────┼─────────────┼─────────────┼────────────────┼───────────────┤
 │  │ .got.plt                    │ 0x3000      │ 0x14        │ 0x3000         │ 0x14          │
 │  ╰─────────────────────────────┴─────────────┴─────────────┴────────────────┴───────────────╯
 │
 │    Segments:
 │  ╭────────┬────────┬────────────────┬───────┬─────────────────────────────╮
 │  │ Start  │ Align  │ Type           │ Perms │ Content                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x34   │ 0x1000 │ program header │ R     │ <program header>            │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x219  │ 0x1    │ interpreter    │ R     │ .interp                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x0    │ 0x1000 │ program        │ R     │ <elf header>                │
 │  │        │        │                │       │ <program header>            │
 │  │        │        │                │       │ .dynstr                     │
 │  │        │        │                │       │ .dynsym                     │
 │  │        │        │                │       │ .hash                       │
 │  │        │        │                │       │ .dynamic                    │
 │  │        │        │                │       │ .rel.plt                    │
 │  │        │        │                │       │ .interp                     │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x1000 │ 0x1000 │ program        │ R     │ .rodata                     │
 │  │        │        │                │       │ .eh_frame                   │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x2000 │ 0x1000 │ program        │ RX    │ .text#0                     │
 │  │        │        │                │       │ .text.__x86.get_pc_thunk.bx │
 │  │        │        │                │       │ .text#1                     │
 │  │        │        │                │       │ .plt                        │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x3000 │ 0x1000 │ program        │ RW    │ .got                        │
 │  │        │        │                │       │ .got.plt                    │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x1a9  │ 0x4    │ dynamic        │ R     │ .dynamic                    │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ -      │ 0x1    │ GNU stack      │ RW    │                             │
 │  ├────────┼────────┼────────────────┼───────┼─────────────────────────────┤
 │  │ 0x3000 │ 0x1    │ GNU relro      │ R     │ .got.plt                    │
 │  ╰────────┴────────┴────────────────┴───────┴─────────────────────────────╯
 │
 │    deduplication facade .comment#1 in hello.o
 │  ╭──────┬──────────────────╮
 │  │ From │ To               │
 │  ├──────┼──────────────────┤
 │  │ 0x0  │ .comment#0 + 0x0 │
 │  ├──────┼──────────────────┤
 │  │ 0x1  │ .comment#0 + 0x1 │
 │  ╰──────┴──────────────────╯
 │
 │    deduplication facade .comment#2 in <plinky>
 │  ╭──────┬───────────────────╮
 │  │ From │ To                │
 │  ├──────┼───────────────────┤
 │  │ 0x0  │ .comment#0 + 0x13 │
 │  ╰──────┴───────────────────╯
 ┴

debug print: built elf
 │
 │  section .dynamic (address: 0x1a9)
 │   │
 │   │  dynamic information | string table: .dynstr
 │   │
 │   │  ╭─────────────────────────┬────────────────╮
 │   │  │ Kind                    │ Value          │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ String table            │ address 0x154  │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ String table size       │ 13 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Symbol table            │ address 0x161  │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Symbol table entry size │ 16 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Hash table              │ address 0x191  │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Jump PLT relocations    │ address 0x209  │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ PLT relocations size    │ 16 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ PLT relocations type    │ Relocations    │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ PLT GOT                 │ address 0x3000 │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Flags                   │ bind_now       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Flags1                  │ pie            │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Null                    │ -              │
 │   │  ╰─────────────────────────┴────────────────╯
 │   ┴
 │
 │    Segments:
 │  ╭───────────────────────────┬───────┬──────────┬─────────────┬──────────┬────────────────┬────────────╮
 │  │ Type                      │ Perms │ Aligment │ File offset │ File len │ Memory address │ Memory len │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Program header table      │ R     │ 0x1000   │ 0x34        │ 0x120    │ 0x34           │ 0x120      │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Interpreter               │ R     │ 0x1      │ 0x219       │ 0x55     │ 0x219          │ 0x55       │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                      │ R     │ 0x1000   │ 0x0         │ 0x26e    │ 0x0            │ 0x26e      │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                      │ R     │ 0x1000   │ 0x1000      │ 0x55     │ 0x1000         │ 0x55       │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                      │ R X   │ 0x1000   │ 0x2000      │ 0x8b     │ 0x2000         │ 0x8b       │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Load                      │ RW    │ 0x1000   │ 0x3000      │ 0x14     │ 0x3000         │ 0x14       │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ Dynamic                   │ R     │ 0x4      │ 0x1a9       │ 0x60     │ 0x1a9          │ 0x60       │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ GNU stack                 │ RW    │ 0x1      │ 0x0         │ 0x0      │ 0x0            │ 0x0        │
 │  ├───────────────────────────┼───────┼──────────┼─────────────┼──────────┼────────────────┼────────────┤
 │  │ GNU relocations read-only │ R     │ 0x1      │ 0x3000      │ 0x14     │ 0x3000         │ 0x14       │
 │  ╰───────────────────────────┴───────┴──────────┴─────────────┴──────────┴────────────────┴────────────╯
 ┴

==============

running exited with exit status: 0

=== stdout ===
Hello world

no stderr present
